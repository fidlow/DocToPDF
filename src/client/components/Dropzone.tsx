import * as React from "react";
import { useState, useEffect } from "react";
import axios, { AxiosRequestConfig } from "axios";
import { ResponseInterface } from "../../common/ResponseInterface";

export default function Dropzone() {
  const validTypes = ["text/xml"];
  const disabled = false;
  const [hightlight, setHiglight] = useState<boolean>(false);
  const [files, setFiles] = useState<File[]>([]);
  const [filesWithoutDuplicates, setFilesWithoutDuplicates] = useState<File[]>(
    []
  );
  const [isUploading, setIsUploading] = useState<boolean>(false);
  const [errorMessage, setErrorMessage] = useState<string>("");
  const fileInputRef = React.useRef(null);
  const openFileDialog = () => {
    fileInputRef.current.click();
  };
  const onFilesAddedByClick = evt => {
    const files: File[] = evt.target.files;
    handleFiles(files);
  };
  const onDragOver = evt => {
    evt.preventDefault();
    setHiglight(true);
  };
  const onDragLeave = e => {
    e.preventDefault();
    setHiglight(false);
  };
  const onDragEnter = e => {
    e.preventDefault();
  };
  const onDrop = e => {
    e.preventDefault();
    setHiglight(false);
    const files = e.dataTransfer.files;
    handleFiles(files);
  };
  const handleFiles = files => {
    const validateFile = file => {
      if (validTypes.indexOf(file.type) === -1) {
        return { isError: true, message: "Не тот тип файла. Нужен .xml" };
      }
      return { isError: false, message: null };
    };
    for (let i = 0; i < files.length; i++) {
      if (!validateFile(files[i]).isError) {
        setFiles(prevArray => [...prevArray, files[i]]);
      }
    }    
  };  
  const fileSize = size => {
    if (size === 0) return "0 Bytes";
    const k = 1024;
    const sizes = ["Байт", "КБ", "МБ", "ГБ", "ТБ"];
    const i = Math.floor(Math.log(size) / Math.log(k));
    return parseFloat((size / Math.pow(k, i)).toFixed(2)) + " " + sizes[i];
  };
  useEffect(() => {
    let filteredArray = files.reduce((file, current) => {
      const x = file.find(item => item.name === current.name);
      if (!x) {
        return file.concat([current]);
      } else {
        return file;
      }
    }, []);
    setFilesWithoutDuplicates([...filteredArray]);
  }, [files]);
  const removeFile = name => {
    setFiles([...files.filter(file => file.name !== name)]);
  };
  const uploadFiles = async () => {
    setIsUploading(true);
    const configUpload: AxiosRequestConfig = {
      responseType: "arraybuffer",
      onUploadProgress: function(progressEvent) {
        // var percentCompleted = Math.round(
        //   (progressEvent.loaded * 100) / progressEvent.total
        // );
      },
    };
    const sendFile = file => {
      return new Promise((resolve, reject) => {
        const formData = new FormData();
        formData.append("file", file);
        axios
          .post<ResponseInterface>("/api/upload", {formData})
          .then(res => {
            if (res.data.error) {
              if ((res.data.message === "XsdValidationError"))
                setErrorMessage("Это не акт сдачи-приемки");
              else setErrorMessage(res.data.message);
            } else {
              axios
                .post<Blob>(
                  "/api/download",
                  { filename: res.data.message },
                  configUpload
                )
                .then(res => {
                  const url = window.URL.createObjectURL(
                    new Blob([res.data], { type: "application/pdf" })
                  );
                  const link = document.createElement("a");
                  link.href = url;
                  link.setAttribute("download", "file.pdf"); //or any other extension
                  document.body.appendChild(link);
                  link.click();
                  resolve();
                });
            }
          })
          // .catch(() => {
          //   console.warn("Upload error on file", file.name);
          // });
      });
    } 
    const promises = [];
    for (let i = 0; i < filesWithoutDuplicates.length; i++) {
      promises.push(sendFile(filesWithoutDuplicates[i]));
    }
    try {
      await Promise.all(promises);
      setIsUploading(false);
      setFiles([]);
    } catch (e) {
      // console.warn("Upload error", e.message );
    }
  };
  return (
    <div className="dropzone">
      {isUploading ? (
        <div>
          {errorMessage !== "" ? (
            <span>
              {errorMessage}. Попробуйте <a href="/">загрузить</a> другой файл.
            </span>
          ) : (
            <span>Идет загрузка...</span>
          )}
        </div>
      ) : (
        <>
          <div
            onClick={openFileDialog}
            onDragEnter={onDragEnter}
            onDragOver={onDragOver}
            onDragLeave={onDragLeave}
            onDrop={onDrop}
            className={`dropzone__container ${
              hightlight ? "dropzone__hightlight" : ""
            }`}
            style={{ cursor: "pointer" }}
          >
            <img
              alt="upload"
              className="dropzone__icon"
              src="images/cloud_upload.svg"
            />
            <input
              ref={fileInputRef}
              className="dropzone__fileinput"
              type="file"
              multiple
              onChange={onFilesAddedByClick}
              accept={validTypes.join(",")}
            />
            <span>Кликните сюда или переместите XML файлы на это поле</span>
          </div>
          <div className="files">
            {filesWithoutDuplicates.map(file => (
              <div className="file" key={file.name + file.size}>
                <div>
                  <span className="file__name">{file.name}</span>
                  <span className="file__size">{fileSize(file.size)}</span>{" "}
                </div>
                <div
                  className="file__remove"
                  onClick={() => removeFile(file.name)}
                >
                  X
                </div>
              </div>
            ))}
          </div>
          {filesWithoutDuplicates.length > 0 && (
            <button id="submit"className="button button--hero" onClick={uploadFiles}>
              Получить PDF
            </button>
          )}
        </>
      )}
    </div>
  );
}
