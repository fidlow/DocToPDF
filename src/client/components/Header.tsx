import * as React from "react";

export default function Header() {
    return <header className="header">
      <div className="logo">
        <p className="logo__title">Doc To Pdf</p>
      </div>
      <nav className="menu">
        <button className="menu__button" type="button">
          <use
            className="menu__icon-menu"
            href="images/sprite.svg#menu"
          ></use>
          <use
            className="menu__icon-cross"
            href="images/sprite.svg#cross"
          ></use>
        </button>
        <ul className="menu__list" id="menu__list">
          <li className="menu__item">
            <a className="menu__link menu__link--current" href="/">
                  Главная
            </a>
          </li>
          <li className="menu__item">
            <a className="menu__link menu__link--inactive">О нас</a>
          </li>
        </ul>
      </nav>
    </header>;
  }