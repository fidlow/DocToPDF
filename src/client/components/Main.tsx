import * as React from "react";
import Dropzone from "./Dropzone";

export default function Main() {
  return (
    <main className="main main--hero">
      {<h1>Загрузите XML акта сдачи-приёмки и получите PDF бесплатно</h1>}
      <Dropzone />
      <p className="main__paragraph main__paragraph--hero"></p>
    </main>
  );
}
