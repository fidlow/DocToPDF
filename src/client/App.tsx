import * as React from "react";
import { render } from "react-dom";
import "./styles/index.css";
import Header from "./components/Header";
import Main from "./components/Main";

const App = () => {
  return (
    <div className="page">
      <Header />
      <Main />       
    </div>
  );
};
export default App;



