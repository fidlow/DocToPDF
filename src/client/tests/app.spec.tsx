import { fireEvent, render, cleanup, waitFor } from "@testing-library/react";
import { configure, shallow, mount } from "enzyme";
import ReactSixteenAdapter from "enzyme-adapter-react-16";
import App from "../App";
import React = require("react");
import Header from "../components/Header";
import Main from "../components/Main";
import Dropzone from "../components/Dropzone";
configure({ adapter: new ReactSixteenAdapter() });
import axios from "axios";
jest.mock("axios");
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe("dropzoneComponent", () => {
  afterEach(cleanup);
  it("renders correctly", () => {
    const component = mount(<App />);
    expect(component).toMatchSnapshot();
  });
  it("renders Header", () => {
    const wrapper = shallow(<Header />);
    const text = "Главная";
    expect(wrapper.contains(text)).toEqual(true);
  });
  it("renders Main", () => {
    const wrapper = shallow(<Main />);
    const isP = !!wrapper.find("p");
    expect(isP).toEqual(true);
  });
  it("renders Dropzone", () => {
    const { getByText } = render(<Dropzone />);
    const firstText = getByText(
      "Кликните сюда или переместите XML файлы на это поле"
    );
    expect(firstText).toBeInTheDocument();
  });
  it("click on Dropzone", async () => {
    const { container } = render(<Dropzone />);
    fireEvent.click(container.querySelector("div .dropzone__container"));
  });

  it("css class changes with drag on Dropzone", async () => {
    const { container } = render(<Dropzone />);
    const element = container.querySelector("div.dropzone__container");
    fireEvent.dragEnter(element);
    expect(
      container.querySelector("div.dropzone__hightlight")
    ).not.toBeInTheDocument();
    fireEvent.dragOver(element);
    expect(
      container.querySelector("div.dropzone__hightlight")
    ).toBeInTheDocument();
    fireEvent.dragLeave(element);
    expect(
      container.querySelector("div.dropzone__hightlight")
    ).not.toBeInTheDocument();
  });
  it("add, delete file on Dropzone by click", async () => {
    const { getByText, queryByText, container } = render(<Dropzone />);
    fireEvent.change(container.querySelector("input.dropzone__fileinput"), {
      target: {
        files: {
          0: { name: "test.xml", type: "text/xml", size: 1024 },
          length: 1,
        },
      },
    });
    expect(
      getByText("test.xml", { selector: "span.file__name" })
    ).toBeInTheDocument();
    expect(
      getByText("1 КБ", { selector: "span.file__size" })
    ).toBeInTheDocument();
    fireEvent.click(container.querySelector("div.file__remove"));
    expect(
      queryByText("test.xml", { selector: "span.file__name" })
    ).not.toBeInTheDocument();
  });
  it("add files with zero size on Dropzone", async () => {
    const { getByText, container } = render(<Dropzone />);
    fireEvent.change(container.querySelector("input.dropzone__fileinput"), {
      target: {
        files: {
          0: { name: "test.xml", type: "text/xml", size: 0 },
          length: 1
        },
      },
    });
    expect(getByText("test.xml")).toBeDefined();
  });
  it("add files with different type on Dropzone", async () => {
    const { queryByText, container } = render(<Dropzone />);
    fireEvent.change(container.querySelector("input.dropzone__fileinput"), {
      target: {
        files: {
          0: { name: "test.xml", type: "text/xml", size: 1024 },
          1: { name: "test.json", type: "application/json", size: 56 },
          length: 2
        },
      },
    });
    expect(queryByText("test.xml")).toBeDefined();
    expect(queryByText("test.json")).toBeNull();
  });
  it("add duplicate files on Dropzone", async () => {
    const { getAllByText, container } = render(<Dropzone />);
    fireEvent.change(container.querySelector("input.dropzone__fileinput"), {
      target: {
        files: {
          0: { name: "test.xml", type: "text/xml", size: 1024 },
          1: { name: "test.xml", type: "text/xml", size: 1024 },
          length: 2,
        },
      },
    });
    expect(getAllByText("test.xml").length).toEqual(1);
  });
  it("add, delete file on Dropzone by drop", async () => {
    const { getByText, queryByText, container } = render(<Dropzone />);
    fireEvent.drop(container.querySelector("div.dropzone__container"), {
      dataTransfer: {
        files: {
          0: { name: "test.xml", type: "text/xml", size: 1024 },
          length: 1,
        },
      },
    });
    expect(
      getByText("test.xml", { selector: "span.file__name" })
    ).toBeInTheDocument();
    expect(
      getByText("1 КБ", { selector: "span.file__size" })
    ).toBeInTheDocument();
    fireEvent.click(container.querySelector("div.file__remove"));
    expect(
      queryByText("test.xml", { selector: "span.file__name" })
    ).not.toBeInTheDocument();
  });
  it("upload", async () => {
    const { container, getByText } = render(<Dropzone />);
    expect(container.querySelector("#submit")).not.toBeInTheDocument();
    fireEvent.change(container.querySelector("input.dropzone__fileinput"), {
      target: {
        files: {
          0: { name: "test.xml", type: "text/xml", size: 1024 },
          length: 1,
        },
      },
    });
    expect(container.querySelector("#submit")).toBeInTheDocument();
    global.URL.createObjectURL = jest.fn();
    mockedAxios.post
      .mockResolvedValueOnce({ data: { error: false, message: "test" } })
      .mockResolvedValueOnce(new File(["test"], "test.pdf"));
    fireEvent.click(container.querySelector("#submit"));
    await waitFor(() =>
      getByText(/Кликните сюда или переместите XML файлы на это поле/)
    );
  });
  it("handle XsdValidationError", async () => {
    const { container, getByText } = render(<Dropzone />);
    expect(container.querySelector("#submit")).not.toBeInTheDocument();
    fireEvent.change(container.querySelector("input.dropzone__fileinput"), {
      target: {
        files: {
          0: { name: "test.xml", type: "text/xml", size: 1024 },
          length: 1,
        },
      },
    });
    expect(container.querySelector("#submit")).toBeInTheDocument();
    mockedAxios.post.mockResolvedValueOnce({
      data: { error: true, message: "XsdValidationError" },
    });
    fireEvent.click(container.querySelector("#submit"));
    await waitFor(() => getByText(/Это не акт сдачи-приемки/));
  });
  it("handle any error", async () => {
    const { container, getByText } = render(<Dropzone />);
    expect(container.querySelector("#submit")).not.toBeInTheDocument();
    fireEvent.change(container.querySelector("input.dropzone__fileinput"), {
      target: {
        files: {
          0: { name: "test.xml", type: "text/xml", size: 1024 },
          length: 1,
        },
      },
    });
    expect(container.querySelector("#submit")).toBeInTheDocument();
    mockedAxios.post.mockResolvedValueOnce({
      data: { error: true, message: "AnyError" },
    });
    fireEvent.click(container.querySelector("#submit"));
    await waitFor(() => getByText(/AnyError/));
  });
  // it("handle unknown error", async () => {
    //   // global.console = {
    //   //   warn: jest.fn(),
    //   //   log: jest.fn(),
    //   //   ...global.console
    //   // }
    //   // const spy = jest.spyOn(global.console, 'warn').mockImplementation()
    //   // const props = {};
    //   // const wrapper = mount(<Dropzone {...props} />);
    //   // expect(wrapper.find('#submit')).toHaveLength(0);
    //   // const input = wrapper.find('input.dropzone__fileinput');
    //   // expect(input).toHaveLength(1);
    //   // input.simulate('change', {
    //   //   target: {
    //   //     files: {
    //   //       0: { name: "test.xml", type: "text/xml", size: 1024 },
    //   //       length: 1
    //   //     },
    //   //   },
    //   // });
    //   // expect(wrapper.find('#submit')).toHaveLength(1);
    //   // mockedAxios.post.mockRejectedValueOnce("test");
    //   // wrapper.find('#submit').simulate('click');

  //   const { container } = render(<Dropzone />);
  //   expect(container.querySelector("#submit")).not.toBeInTheDocument();
  //   fireEvent.change(container.querySelector("input.dropzone__fileinput"), {
  //     target: {
  //       files: {
  //         0: { name: "test.xml", type: "text/xml", size: 1024 },
  //         length: 1,
  //       },
  //     },
  //   });
  //   expect(container.querySelector("#submit")).toBeInTheDocument();
  //   mockedAxios.post.mockRejectedValueOnce("test2");
  //   fireEvent.click(container.querySelector("#submit"));
  //   // expect(console.warn).toBeCalled();
  // });
});
