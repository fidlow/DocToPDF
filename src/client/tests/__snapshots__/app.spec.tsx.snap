// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dropzoneComponent renders correctly 1`] = `
<App>
  <div
    className="page"
  >
    <Header>
      <header
        className="header"
      >
        <div
          className="logo"
        >
          <p
            className="logo__title"
          >
            Doc To Pdf
          </p>
        </div>
        <nav
          className="menu"
        >
          <button
            className="menu__button"
            type="button"
          >
            <use
              className="menu__icon-menu"
              href="images/sprite.svg#menu"
            />
            <use
              className="menu__icon-cross"
              href="images/sprite.svg#cross"
            />
          </button>
          <ul
            className="menu__list"
            id="menu__list"
          >
            <li
              className="menu__item"
            >
              <a
                className="menu__link menu__link--current"
                href="/"
              >
                Главная
              </a>
            </li>
            <li
              className="menu__item"
            >
              <a
                className="menu__link menu__link--inactive"
              >
                О нас
              </a>
            </li>
          </ul>
        </nav>
      </header>
    </Header>
    <Main>
      <main
        className="main main--hero"
      >
        <h1>
          Загрузите XML акта сдачи-приёмки и получите PDF бесплатно
        </h1>
        <Dropzone>
          <div
            className="dropzone"
          >
            <div
              className="dropzone__container "
              onClick={[Function]}
              onDragEnter={[Function]}
              onDragLeave={[Function]}
              onDragOver={[Function]}
              onDrop={[Function]}
              style={
                Object {
                  "cursor": "pointer",
                }
              }
            >
              <img
                alt="upload"
                className="dropzone__icon"
                src="images/cloud_upload.svg"
              />
              <input
                accept="text/xml"
                className="dropzone__fileinput"
                multiple={true}
                onChange={[Function]}
                type="file"
              />
              <span>
                Кликните сюда или переместите XML файлы на это поле
              </span>
            </div>
            <div
              className="files"
            />
          </div>
        </Dropzone>
        <p
          className="main__paragraph main__paragraph--hero"
        />
      </main>
    </Main>
  </div>
</App>
`;
